DisambiguationAuthenticator is re-prompting for login information

@1gravity @ThomasSHA

One more thing: 
When I use the "loginAndBuildClient"-Login-Method -> login -> and then wait for an hour (until the token is no longer valid) inside the app and then try to do e.g. a list directory, it does not work. Like before the fix. You could tell me now, I should login again, or every hour, or every time I try to make a file operation, but I think that's now how it should work (please correct me if I'm wrong) :).

Following the code after the Token seems to be expired and I try to list of the root folder. (It seems like Token cannot be refreshed):

```
03-29 12:09:00.553 27475-29394/com.<PRIVATE>.onedrive D/DefaultExecutors[performOnBackground] - 67: Starting background task, current active count: 0
03-29 12:09:00.573 27475-3742/com.<PRIVATE>.onedrive D/AuthorizationInterceptor[intercept] - 71: Intercepting request, https://<PRIVATE><PRIVATE>-my.sharepoint.com/_api/v2.0/me/drive/root?expand=children
03-29 12:09:00.574 27475-3742/com.<PRIVATE>.onedrive D/AuthorizationInterceptor[intercept] - 82: Found account information
03-29 12:09:00.579 27475-3742/com.<PRIVATE>.onedrive V/TokenCacheItem: 2016-03-29 10:09:00-5ee7d0d8-ce32-42ea-bb09-476d4825a918-expiresOn:Tue Mar 29 11:27:27 GMT+02:00 2016 timeWithBuffer:Tue Mar 29 12:14:00 GMT+02:00 2016 Buffer:300 ver:1.1.7
03-29 12:09:00.585 27475-3742/com.<PRIVATE>.onedrive D/AuthorizationInterceptor[intercept] - 84: Account access token is expired, refreshing
03-29 12:09:00.587 27475-3742/com.<PRIVATE>.onedrive D/ADALAccountInfo[refresh] - 112: Refreshing access token...
03-29 12:09:00.587 27475-3742/com.<PRIVATE>.onedrive D/ADALAuthenticator[loginSilent] - 380: Starting login silent
03-29 12:09:00.588 27475-3742/com.<PRIVATE>.onedrive V/AuthenticationContext: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Sending async task from thread:3742 ver:1.1.7
03-29 12:09:00.590 27475-27679/com.<PRIVATE>.onedrive V/AuthenticationContext: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Running task in thread:27679 ver:1.1.7
03-29 12:09:00.595 27475-27679/com.<PRIVATE>.onedrive V/AuthenticationContext: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Token request started ver:1.1.7
03-29 12:09:00.601 27475-27679/com.<PRIVATE>.onedrive V/AuthenticationContext: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Checking refresh tokens ver:1.1.7
03-29 12:09:00.603 27475-27679/com.<PRIVATE>.onedrive V/AuthenticationContext: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Looking for regular refresh token ver:1.1.7
03-29 12:09:00.605 27475-27679/com.<PRIVATE>.onedrive V/AuthenticationContext: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Looking for Multi Resource Refresh token ver:1.1.7
03-29 12:09:00.607 27475-27679/com.<PRIVATE>.onedrive V/StorageHelper: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Starting decryption ver:1.1.7
03-29 12:09:00.613 27475-27679/com.<PRIVATE>.onedrive V/StorageHelper: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Finished decryption ver:1.1.7
03-29 12:09:00.616 27475-27679/com.<PRIVATE>.onedrive V/AuthenticationContext: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Refresh token is available and id:x/kcraqUBQSC27Gs7qUY8Kxe4Neni5ziKUrfeJGS4Uo= Key used:https://login.windows.net/common$null$<PRIVATE_CLIENT_ID>$y$1a4067c8-099e-4c54-8b1a-3893b2601219 ver:1.1.7
03-29 12:09:00.617 27475-27679/com.<PRIVATE>.onedrive V/AuthenticationContext: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Refresh token is available and it will attempt to refresh token ver:1.1.7
03-29 12:09:00.619 27475-27679/com.<PRIVATE>.onedrive V/AuthenticationContext: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Process refreshToken for Request authority:https://login.windows.net/common resource:https://<PRIVATE><PRIVATE>-my.sharepoint.com/_api/v2.0/me clientid:<PRIVATE_CLIENT_ID> refreshTokenId:x/kcraqUBQSC27Gs7qUY8Kxe4Neni5ziKUrfeJGS4Uo= ver:1.1.7
03-29 12:09:00.622 27475-27679/com.<PRIVATE>.onedrive V/WebRequestHandler: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-WebRequestHandler thread27679 ver:1.1.7
03-29 12:09:00.623 27475-27679/com.<PRIVATE>.onedrive V/HttpWebRequest: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-HttpWebRequest send thread:27679 ver:1.1.7
03-29 12:09:00.625 27475-27679/com.<PRIVATE>.onedrive V/HttpWebRequest: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-HttpWebRequest setupConnection thread:27679 ver:1.1.7
03-29 12:09:00.627 27475-27679/com.<PRIVATE>.onedrive V/HttpWebRequest: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Setting header: x-client-Ver ver:1.1.7
03-29 12:09:00.629 27475-27679/com.<PRIVATE>.onedrive V/HttpWebRequest: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Setting header: x-client-last-request ver:1.1.7
03-29 12:09:00.631 27475-27679/com.<PRIVATE>.onedrive V/HttpWebRequest: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Setting header: x-client-last-response-time ver:1.1.7
03-29 12:09:00.632 27475-27679/com.<PRIVATE>.onedrive V/HttpWebRequest: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Setting header: Host ver:1.1.7
03-29 12:09:00.633 27475-27679/com.<PRIVATE>.onedrive V/HttpWebRequest: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Setting header: x-client-last-endpoint ver:1.1.7
03-29 12:09:00.634 27475-27679/com.<PRIVATE>.onedrive V/HttpWebRequest: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Setting header: x-client-DM ver:1.1.7
03-29 12:09:00.636 27475-27679/com.<PRIVATE>.onedrive V/HttpWebRequest: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Setting header: x-client-OS ver:1.1.7
03-29 12:09:00.637 27475-27679/com.<PRIVATE>.onedrive V/HttpWebRequest: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Setting header: Accept ver:1.1.7
03-29 12:09:00.638 27475-27679/com.<PRIVATE>.onedrive V/HttpWebRequest: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Setting header: x-client-SKU ver:1.1.7
03-29 12:09:00.639 27475-27679/com.<PRIVATE>.onedrive V/HttpWebRequest: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Setting header: client-request-id ver:1.1.7
03-29 12:09:00.640 27475-27679/com.<PRIVATE>.onedrive V/HttpWebRequest: 2016-03-29 10:09:00-fba516de-83d5-4059-aa28-75d394093c3b-Setting header: x-ms-PKeyAuth ver:1.1.7
03-29 12:09:01.570 27475-27679/com.<PRIVATE>.onedrive E/HttpWebRequest: SERVER_ERROR:2016-03-29 10:09:01-fba516de-83d5-4059-aa28-75d394093c3b-IOException:https://login.windows.net/common/oauth2/token ver:1.1.7 
03-29 12:09:01.572 27475-27679/com.<PRIVATE>.onedrive V/HttpWebRequest: 2016-03-29 10:09:01-fba516de-83d5-4059-aa28-75d394093c3b-Status code:400 ver:1.1.7
03-29 12:09:01.577 27475-27679/com.<PRIVATE>.onedrive V/HttpWebRequest: 2016-03-29 10:09:01-fba516de-83d5-4059-aa28-75d394093c3b-Response is received ver:1.1.7
03-29 12:09:01.579 27475-27679/com.<PRIVATE>.onedrive V/Oauth: 2016-03-29 10:09:01-fba516de-83d5-4059-aa28-75d394093c3b-Token request does not have exception ver:1.1.7
                                                            Timestamp: 2016-03-29 10:09:06Z ver:1.1.7
03-29 12:09:01.583 27475-27679/com.<PRIVATE>.onedrive V/Oauth: 2016-03-29 10:09:01-3ce6244c-626c-4b39-9c27-ef0449a085ca-Response correlationId:3ce6244c-626c-4b39-9c27-ef0449a085ca ver:1.1.7
03-29 12:09:01.587 27475-27679/com.<PRIVATE>.onedrive V/AuthenticationContext: 2016-03-29 10:09:01-3ce6244c-626c-4b39-9c27-ef0449a085ca-Refresh token is not returned or empty ver:1.1.7
                                                                            Timestamp: 2016-03-29 10:09:06Z
03-29 12:09:01.592 27475-27679/com.<PRIVATE>.onedrive V/AuthenticationContext: 2016-03-29 10:09:01-3ce6244c-626c-4b39-9c27-ef0449a085ca-Remove refresh item from cache:https://login.windows.net/common$null$<PRIVATE_CLIENT_ID>$y$1a4067c8-099e-4c54-8b1a-3893b2601219 ver:1.1.7
03-29 12:09:01.593 27475-27679/com.<PRIVATE>.onedrive V/AuthenticationContext: 2016-03-29 10:09:01-3ce6244c-626c-4b39-9c27-ef0449a085ca-Token request started ver:1.1.7
03-29 12:09:01.597 27475-27679/com.<PRIVATE>.onedrive V/AuthenticationContext: 2016-03-29 10:09:01-3ce6244c-626c-4b39-9c27-ef0449a085ca-Checking refresh tokens ver:1.1.7
03-29 12:09:01.599 27475-27679/com.<PRIVATE>.onedrive V/AuthenticationContext: 2016-03-29 10:09:01-3ce6244c-626c-4b39-9c27-ef0449a085ca-Looking for regular refresh token ver:1.1.7
03-29 12:09:01.601 27475-27679/com.<PRIVATE>.onedrive V/AuthenticationContext: 2016-03-29 10:09:01-3ce6244c-626c-4b39-9c27-ef0449a085ca-Looking for Multi Resource Refresh token ver:1.1.7
03-29 12:09:01.603 27475-27679/com.<PRIVATE>.onedrive V/AuthenticationContext: 2016-03-29 10:09:01-3ce6244c-626c-4b39-9c27-ef0449a085ca-Refresh token is not available ver:1.1.7
03-29 12:09:01.606 27475-27679/com.<PRIVATE>.onedrive E/AuthenticationContext: AUTH_REFRESH_FAILED_PROMPT_NOT_ALLOWED:2016-03-29 10:09:01-3ce6244c-626c-4b39-9c27-ef0449a085ca-Prompt is not allowed and failed to get token: ver:1.1.7 
03-29 12:09:01.610 27475-27475/com.<PRIVATE>.onedrive D/ADALAuthenticator$3[onError] - 417: Silent authentication failure from ADAL; Code Refresh token is failed and prompt is not allowed
03-29 12:09:01.612 27475-3742/com.<PRIVATE>.onedrive E/DefaultHttpProvider[sendRequestInternal] - 303: Error during http request
                                                                                                    com.onedrive.sdk.core.ClientException: Error during http request
                                                                                                        at com.onedrive.sdk.http.DefaultHttpProvider.sendRequestInternal(DefaultHttpProvider.java:300)
                                                                                                        at com.onedrive.sdk.http.DefaultHttpProvider.access$000(DefaultHttpProvider.java:48)
                                                                                                        at com.onedrive.sdk.http.DefaultHttpProvider$1.run(DefaultHttpProvider.java:127)
                                                                                                        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
                                                                                                        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
                                                                                                        at java.lang.Thread.run(Thread.java:818)
                                                                                                     Caused by: com.onedrive.sdk.authentication.ClientAuthenticatorException: Silent authentication failure from ADAL; Code Refresh token is failed and prompt is not allowed
                                                                                                        at com.onedrive.sdk.authentication.ADALAuthenticator$3.onError(ADALAuthenticator.java:418)
                                                                                                        at com.microsoft.aad.adal.AuthenticationContext$CallbackHandler$1.run(AuthenticationContext.java:1034)
                                                                                                        at android.os.Handler.handleCallback(Handler.java:739)
                                                                                                        at android.os.Handler.dispatchMessage(Handler.java:95)
                                                                                                        at android.os.Looper.loop(Looper.java:148)
                                                                                                        at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                                                                        at java.lang.reflect.Method.invoke(Native Method)
                                                                                                        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                                                                        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
                                                                                                     Caused by: com.microsoft.aad.adal.AuthenticationException: Refresh token is failed and prompt is not allowed
                                                                                                        at com.microsoft.aad.adal.AuthenticationContext.localFlow(AuthenticationContext.java:1294)
                                                                                                        at com.microsoft.aad.adal.AuthenticationContext.acquireTokenAfterValidation(AuthenticationContext.java:1229)
                                                                                                        at com.microsoft.aad.adal.AuthenticationContext.acquireTokenLocalCall(AuthenticationContext.java:1123)
                                                                                                        at com.microsoft.aad.adal.AuthenticationContext.refreshToken(AuthenticationContext.java:1609)
                                                                                                        at com.microsoft.aad.adal.AuthenticationContext.localFlow(AuthenticationContext.java:1261)
                                                                                                        at com.microsoft.aad.adal.AuthenticationContext.acquireTokenAfterValidation(AuthenticationContext.java:1229)
                                                                                                        at com.microsoft.aad.adal.AuthenticationContext.acquireTokenLocalCall(AuthenticationContext.java:1123)
                                                                                                        at com.microsoft.aad.adal.AuthenticationContext.access$600(AuthenticationContext.java:58)
                                                                                                        at com.microsoft.aad.adal.AuthenticationContext$4.call(AuthenticationContext.java:1072)
                                                                                                        at com.microsoft.aad.adal.AuthenticationContext$4.call(AuthenticationContext.java:1067)
                                                                                                        at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                                                        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
                                                                                                        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
                                                                                                        at java.lang.Thread.run(Thread.java:818)
03-29 12:09:01.613 27475-3742/com.<PRIVATE>.onedrive D/DefaultExecutors[performOnForeground] - 125: Starting foreground task, current active count:0, with exception com.onedrive.sdk.core.ClientException: Error during http request
03-29 12:09:01.615 27475-27475/com.<PRIVATE>.onedrive E/MainActivity: listException 
                                                                   com.onedrive.sdk.core.ClientException: Error during http request
                                                                       at com.onedrive.sdk.http.DefaultHttpProvider.sendRequestInternal(DefaultHttpProvider.java:300)
                                                                       at com.onedrive.sdk.http.DefaultHttpProvider.access$000(DefaultHttpProvider.java:48)
                                                                       at com.onedrive.sdk.http.DefaultHttpProvider$1.run(DefaultHttpProvider.java:127)
                                                                       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
                                                                       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
                                                                       at java.lang.Thread.run(Thread.java:818)
                                                                    Caused by: com.onedrive.sdk.authentication.ClientAuthenticatorException: Silent authentication failure from ADAL; Code Refresh token is failed and prompt is not allowed
                                                                       at com.onedrive.sdk.authentication.ADALAuthenticator$3.onError(ADALAuthenticator.java:418)
                                                                       at com.microsoft.aad.adal.AuthenticationContext$CallbackHandler$1.run(AuthenticationContext.java:1034)
                                                                       at android.os.Handler.handleCallback(Handler.java:739)
                                                                       at android.os.Handler.dispatchMessage(Handler.java:95)
                                                                       at android.os.Looper.loop(Looper.java:148)
                                                                       at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                                       at java.lang.reflect.Method.invoke(Native Method)
                                                                       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                                       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
                                                                    Caused by: com.microsoft.aad.adal.AuthenticationException: Refresh token is failed and prompt is not allowed
                                                                       at com.microsoft.aad.adal.AuthenticationContext.localFlow(AuthenticationContext.java:1294)
                                                                       at com.microsoft.aad.adal.AuthenticationContext.acquireTokenAfterValidation(AuthenticationContext.java:1229)
                                                                       at com.microsoft.aad.adal.AuthenticationContext.acquireTokenLocalCall(AuthenticationContext.java:1123)
                                                                       at com.microsoft.aad.adal.AuthenticationContext.refreshToken(AuthenticationContext.java:1609)
                                                                       at com.microsoft.aad.adal.AuthenticationContext.localFlow(AuthenticationContext.java:1261)
                                                                       at com.microsoft.aad.adal.AuthenticationContext.acquireTokenAfterValidation(AuthenticationContext.java:1229)
                                                                       at com.microsoft.aad.adal.AuthenticationContext.acquireTokenLocalCall(AuthenticationContext.java:1123)
                                                                       at com.microsoft.aad.adal.AuthenticationContext.access$600(AuthenticationContext.java:58)
                                                                       at com.microsoft.aad.adal.AuthenticationContext$4.call(AuthenticationContext.java:1072)
                                                                       at com.microsoft.aad.adal.AuthenticationContext$4.call(AuthenticationContext.java:1067)
                                                                       at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
                                                                       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
                                                                       at java.lang.Thread.run(Thread.java:818)

```

--- a/gradle.properties
+++ b/gradle.properties
@@ -23,5 +23,5 @@ mavenGroupId         = com.onedrive.sdk
 mavenArtifactId      = onedrive-sdk-android
 mavenMajorVersion    = 1
 mavenMinorVersion    = 1
-mavenPatchVersion    = 3
+mavenPatchVersion    = 4
 nightliesUrl         = http://dl.bintray.com/onedrive/Maven

--- a/onedrivesdk/src/main/java/com/onedrive/sdk/authentication/ADALAuthenticator.java
+++ b/onedrivesdk/src/main/java/com/onedrive/sdk/authentication/ADALAuthenticator.java
@@ -421,7 +421,8 @@ public abstract class ADALAuthenticator implements IAuthenticator {
                 loginSilentWaiter.signal();
             }
         };
-        mAdalContext.acquireTokenSilent(mResourceUrl.get(), getClientId(), mUserId.get(), callback);
+
+        mAdalContext.acquireTokenSilent(mOneDriveServiceInfo.get().serviceResourceId, getClientId(), mUserId.get(), callback);
 
         loginSilentWaiter.waitForSignal();
         //noinspection ThrowableResultOfMethodCallIgnored

--- a/onedrivesdk/src/main/java/com/onedrive/sdk/authentication/DisambiguationAuthenticator.java
+++ b/onedrivesdk/src/main/java/com/onedrive/sdk/authentication/DisambiguationAuthenticator.java
@@ -333,7 +333,9 @@ public class DisambiguationAuthenticator implements IAuthenticator {
         if (mMSAAuthenticator.getAccountInfo() != null) {
             mLogger.logDebug("Starting logout of MSA account");
             mMSAAuthenticator.logout();
-        } else if (mADALAuthenticator.getAccountInfo() != null) {
+        }
+
+        if (mADALAuthenticator.getAccountInfo() != null) {
             mLogger.logDebug("Starting logout of ADAL account");
             mADALAuthenticator.logout();
         }

--- a/onedrivesdk/src/main/java/com/onedrive/sdk/authentication/DisambiguationAuthenticator.java
+++ b/onedrivesdk/src/main/java/com/onedrive/sdk/authentication/DisambiguationAuthenticator.java
@@ -22,6 +22,7 @@
 
 package com.onedrive.sdk.authentication;
 
+import com.microsoft.onedrivesdk.BuildConfig;
 import com.onedrive.sdk.concurrency.ICallback;
 import com.onedrive.sdk.concurrency.IExecutors;
 import com.onedrive.sdk.concurrency.SimpleWaiter;
@@ -31,6 +32,9 @@ import com.onedrive.sdk.http.IHttpProvider;
 import com.onedrive.sdk.logger.ILogger;
 
 import android.app.Activity;
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.support.annotation.Nullable;
 
 import java.security.InvalidParameterException;
 import java.util.concurrent.atomic.AtomicReference;
@@ -39,6 +43,21 @@ import java.util.concurrent.atomic.AtomicReference;
  * Wrapper around the ADAL and MSA authenticators which can disambiguate between them.
  */
 public class DisambiguationAuthenticator implements IAuthenticator {
+
+    /**
+     * The preferences for this authenticator.
+     */
+    private static final String DISAMBIGUATION_AUTHENTICATOR_PREFS = "DisambiguationAuthenticatorPrefs";
+
+    /**
+     * The key for the version code
+     */
+    public static final String VERSION_CODE_KEY = "versionCode";
+
+    /**
+     * The key for the account type
+     */
+    public static final String ACCOUNT_TYPE_KEY = "accountType";
 
     /**
      * The current account info object.
@@ -173,31 +192,41 @@ public class DisambiguationAuthenticator implements IAuthenticator {
             }
         };
 
-        mLogger.logDebug("Creating disambiguation ui, waiting for user to sign in");
-        new DisambiguationRequest(mActivity, disambiguationCallback, mLogger).execute();
-        disambiguationWaiter.waitForSignal();
+        AccountType accountType = getAccountTypeInPreferences();
+        String accountName = null;
+        if (accountType != null) {
+            mLogger.logDebug(String.format("Found saved account information %s type of account", accountType));
+        } else {
+            mLogger.logDebug("Creating disambiguation ui, waiting for user to sign in");
+            new DisambiguationRequest(mActivity, disambiguationCallback, mLogger).execute();
+            disambiguationWaiter.waitForSignal();
 
-        //noinspection ThrowableResultOfMethodCallIgnored
-        if (error.get() != null) {
-            throw error.get();
+            //noinspection ThrowableResultOfMethodCallIgnored
+            if (error.get() != null) {
+                throw error.get();
+            }
+            final DisambiguationResponse disambiguationResponse = response.get();
+            accountType = disambiguationResponse.getAccountType();
+            accountName = disambiguationResponse.getAccount();
         }
-        final DisambiguationResponse disambiguationResponse = response.get();
+
         final IAccountInfo accountInfo;
-        switch (disambiguationResponse.getAccountType()) {
+        switch (accountType) {
             case ActiveDirectory:
-                accountInfo = mADALAuthenticator.login(disambiguationResponse.getAccount());
+                accountInfo = mADALAuthenticator.login(accountName);
                 break;
             case MicrosoftAccount:
-                accountInfo = mMSAAuthenticator.login(disambiguationResponse.getAccount());
+                accountInfo = mMSAAuthenticator.login(accountName);
                 break;
             default:
                 final UnsupportedOperationException unsupportedOperationException
                     = new UnsupportedOperationException("Unrecognized account type "
-                                                        + disambiguationResponse.getAccountType());
+                                                        + accountType);
                 mLogger.logError("Unrecognized account type", unsupportedOperationException);
                 throw unsupportedOperationException;
         }
 
+        setAccountTypeInPreferences(accountType);
         mAccountInfo.set(accountInfo);
         return mAccountInfo.get();
     }
@@ -241,10 +270,17 @@ public class DisambiguationAuthenticator implements IAuthenticator {
         }
 
         mLogger.logDebug("Starting login silent");
+
+        final AccountType accountType = getAccountTypeInPreferences();
+        if (accountType != null) {
+            mLogger.logDebug(String.format("Expecting %s type of account", accountType));
+        }
+
         mLogger.logDebug("Checking MSA");
         IAccountInfo accountInfo = mMSAAuthenticator.loginSilent();
         if (accountInfo != null) {
             mLogger.logDebug("Found account info in MSA");
+            setAccountTypeInPreferences(accountType);
             mAccountInfo.set(accountInfo);
             return accountInfo;
         }
@@ -254,8 +290,9 @@ public class DisambiguationAuthenticator implements IAuthenticator {
         mAccountInfo.set(accountInfo);
         if (accountInfo != null) {
             mLogger.logDebug("Found account info in ADAL");
+            setAccountTypeInPreferences(accountType);
         }
-        return accountInfo;
+        return mAccountInfo.get();
     }
 
     /**
@@ -277,7 +314,7 @@ public class DisambiguationAuthenticator implements IAuthenticator {
             @Override
             public void run() {
                 logout();
-                mExecutors.performOnForeground((Void)null, logoutCallback);
+                mExecutors.performOnForeground((Void) null, logoutCallback);
             }
         });
     }
@@ -300,6 +337,12 @@ public class DisambiguationAuthenticator implements IAuthenticator {
             mLogger.logDebug("Starting logout of ADAL account");
             mADALAuthenticator.logout();
         }
+
+        getSharedPreferences()
+                .edit()
+                .clear()
+                .putInt(VERSION_CODE_KEY, BuildConfig.VERSION_CODE)
+                .commit();
     }
 
     /**
@@ -309,5 +352,42 @@ public class DisambiguationAuthenticator implements IAuthenticator {
     @Override
     public IAccountInfo getAccountInfo() {
         return mAccountInfo.get();
+    }
+
+    /**
+     * Gets the shared preferences for this authenticator.
+     * @return The shared preferences.
+     */
+    private SharedPreferences getSharedPreferences() {
+        return mActivity.getSharedPreferences(DISAMBIGUATION_AUTHENTICATOR_PREFS, Context.MODE_PRIVATE);
+    }
+
+    /**
+     * Sets the AccountType from SharedPreferences
+     * @param accountType The account type, can be null
+     */
+    private void setAccountTypeInPreferences(@Nullable final AccountType accountType) {
+        if (accountType == null) {
+            return;
+        }
+
+        getSharedPreferences()
+                .edit()
+                .putString(ACCOUNT_TYPE_KEY, accountType.toString())
+                .putInt(VERSION_CODE_KEY, BuildConfig.VERSION_CODE)
+                .commit();
+    }
+
+    /**
+     * Gets the AccountType from SharedPreferences
+     * @return The account type, null if none was found
+     */
+    @Nullable
+    private AccountType getAccountTypeInPreferences() {
+        final String accountTypeString = getSharedPreferences().getString(ACCOUNT_TYPE_KEY, null);
+        if (accountTypeString == null) {
+            return null;
+        }
+        return AccountType.valueOf(accountTypeString);
     }
 }

